File: scripts/utils/cairo_str_encoding.py
def string_to_felt_array(s, chunk_size):
    """Convert a string into an array of felt values."""
    felt_array = []

    # Split the string into chunks
    chunks = [s[i:i + chunk_size] for i in range(0, len(s), chunk_size)]

    # Convert each chunk into a felt value
    for chunk in chunks:
        felt_value = 0
        for i, char in enumerate(chunk):
            felt_value += ord(char) * (256 ** i)
        felt_array.append(felt_value)

    return felt_array



def felt_array_to_string(felt_array):
    """Convert an array of felt values back to the original string."""
    reconstructed_string = ""
    for felt_value in felt_array:
        while felt_value > 0:
            # Extract the least significant byte (character in ASCII)
            ascii_value = felt_value % 256
            reconstructed_string += chr(ascii_value)
            # Move to the next byte
            felt_value //= 256
    return reconstructed_string


# Example usage
string = 'data:,{"p":"stark-20","op":"mint","tick":"STRK","amt":"1000"}'
encoded_data = string_to_felt_array(string, 31)
# Convert the felt array back to the original string
decoded_data = felt_array_to_string(encoded_data)
# Print the original string
print(decoded_data)

File: scripts/utils/cairo_str_encoding.js
function stringToFeltArray(s, chunkSize) {
    let feltArray = [];
    
    // Function to convert a character to its ASCII value
    function charToAscii(char) {
        return char.charCodeAt(0);
    }

    // Split the string into chunks and convert each chunk to a felt value
    for (let i = 0; i < s.length; i += chunkSize) {
        let chunk = s.substring(i, Math.min(i + chunkSize, s.length));
        let feltValue = BigInt(0);
        for (let j = 0; j < chunk.length; j++) {
            feltValue += BigInt(charToAscii(chunk[j])) * (BigInt(256) ** BigInt(j));
        }
        feltArray.push(feltValue);
    }

    return feltArray;
}

function feltArrayToString(feltArray) {
    let reconstructedString = "";

    // Convert each felt value back to a string chunk
    feltArray.forEach(feltValue => {
        while (feltValue > BigInt(0)) {
            let asciiValue = feltValue % BigInt(256);
            reconstructedString += String.fromCharCode(Number(asciiValue));
            feltValue /= BigInt(256);
        }
    });

    return reconstructedString;
}

// Example usage
let string = 'data:,{"p":"stark-20","op":"mint","tick":"STRK","amt":"1000"}';
console.log(`DATA TO ENCODE: ${string}`); // Prints the original string
let encodedData = stringToFeltArray(string, 31);
console.log(`ENCODED DATA: ${encodedData}`); // Prints the encoded felt array
let decodedData = feltArrayToString(encodedData);

console.log(`DECODED DATA: ${decodedData}`); // Prints the decoded string